File: ts/src/refdict.ts
/**
 * RefDict: Map-like access to git branches or tags.
 */

import git from 'isomorphic-git';
import { PermissionError, type FsModule, type ReflogEntry } from './types.js';
import { validateRefName } from './paths.js';
import { withRepoLock } from './lock.js';
import { readReflog, writeReflogEntry, ZERO_SHA } from './reflog.js';
import { FS } from './fs.js';
import type { GitStore } from './gitstore.js';

/**
 * Map-like access to branches or tags.
 *
 * `store.branches` and `store.tags` are both `RefDict` instances.
 * Supports `get`, `set`, `delete`, `has`, `list`, and async iteration.
 */
export class RefDict {
  private _store: GitStore;
  private _prefix: string; // "refs/heads/" or "refs/tags/"

  constructor(store: GitStore, prefix: string) {
    this._store = store;
    this._prefix = prefix;
  }

  private get _isTags(): boolean {
    return this._prefix === 'refs/tags/';
  }

  private get _fsModule(): FsModule {
    return this._store._fsModule;
  }

  private get _gitdir(): string {
    return this._store._gitdir;
  }

  private _refName(name: string): string {
    return `${this._prefix}${name}`;
  }

  toString(): string {
    const kind = this._isTags ? 'tags' : 'branches';
    return `RefDict('${kind}')`;
  }

  /**
   * Get a branch or tag as an FS snapshot.
   *
   * Branches return a writable FS; tags return a read-only FS.
   *
   * @param name - Branch or tag name.
   * @returns FS snapshot for the ref.
   * @throws {Error} If the branch or tag does not exist.
   */
  async get(name: string): Promise<FS> {
    const refName = this._refName(name);
    let oid: string;
    try {
      oid = await git.resolveRef({ fs: this._fsModule, gitdir: this._gitdir, ref: refName });
    } catch {
      throw new Error(`Key not found: ${name}`);
    }

    if (this._isTags) {
      // Tags may point to annotated tag objects — peel to commit
      const { commit } = await git.readCommit({
        fs: this._fsModule,
        gitdir: this._gitdir,
        oid,
      });
      // oid is already the commit oid after resolveRef for lightweight tags.
      // For annotated tags, readCommit may fail — try reading as tag first.
      return FS._fromCommit(this._store, oid, name, false);
    }

    return FS._fromCommit(this._store, oid, name, true);
  }

  /**
   * Set or create a branch pointing to an FS snapshot's commit.
   *
   * The FS must belong to the same repository. Tags cannot be overwritten;
   * an error is thrown if the tag already exists.
   *
   * @param name - Branch or tag name.
   * @param fs - FS snapshot whose commit to point the ref at.
   * @throws {TypeError} If fs is not an FS instance.
   * @throws {Error} If the FS belongs to a different repository.
   * @throws {Error} If the tag already exists (tags only).
   */
  async set(name: string, fs: FS): Promise<void> {
    validateRefName(name);
    if (!(fs instanceof FS)) throw new TypeError(`Expected FS, got ${typeof fs}`);

    const selfPath = this._fsModule.realpathSync(this._gitdir);
    const fsPath = this._fsModule.realpathSync(fs._gitdir);
    if (selfPath !== fsPath) {
      throw new Error('FS belongs to a different repository');
    }

    const refName = this._refName(name);
    const sig = this._store._signature;
    const committerStr = `${sig.name} <${sig.email}>`;

    await withRepoLock(this._fsModule, this._gitdir, async () => {
      let exists = false;
      try {
        await git.resolveRef({ fs: this._fsModule, gitdir: this._gitdir, ref: refName });
        exists = true;
      } catch { /* not found */ }

      if (exists && this._isTags) {
        throw new Error(`Tag '${name}' already exists`);
      }

      const oldSha = exists
        ? await git.resolveRef({ fs: this._fsModule, gitdir: this._gitdir, ref: refName }).catch(() => ZERO_SHA)
        : ZERO_SHA;

      await git.writeRef({
        fs: this._fsModule,
        gitdir: this._gitdir,
        ref: refName,
        value: fs._commitOid,
        force: true,
      });

      // Write reflog
      const msg = exists
        ? `branch: set to ${(await fs.getMessage()).split('\n')[0]}`
        : `branch: Created from ${(await fs.getMessage()).split('\n')[0]}`;

      await writeReflogEntry(
        this._fsModule,
        this._gitdir,
        refName,
        typeof oldSha === 'string' ? oldSha : ZERO_SHA,
        fs._commitOid,
        committerStr,
        msg,
      );
    });
  }

  /**
   * Set branch to an FS snapshot and return a writable FS bound to it.
   *
   * Convenience method that combines `set()` and `get()`.
   *
   * @param name - Branch name.
   * @param fs - FS snapshot to set (can be read-only).
   * @returns New writable FS bound to the branch.
   */
  async setAndGet(name: string, fs: FS): Promise<FS> {
    await this.set(name, fs);
    return this.get(name);
  }

  /**
   * Delete a branch or tag.
   *
   * @param name - Branch or tag name.
   * @throws {Error} If the branch or tag does not exist.
   */
  async delete(name: string): Promise<void> {
    const refName = this._refName(name);
    await withRepoLock(this._fsModule, this._gitdir, async () => {
      try {
        await git.resolveRef({ fs: this._fsModule, gitdir: this._gitdir, ref: refName });
      } catch {
        throw new Error(`Key not found: ${name}`);
      }
      await git.deleteRef({ fs: this._fsModule, gitdir: this._gitdir, ref: refName });
    });
  }

  /**
   * Check if a branch or tag exists.
   *
   * @param name - Branch or tag name.
   * @returns True if the ref exists.
   */
  async has(name: string): Promise<boolean> {
    const refName = this._refName(name);
    try {
      await git.resolveRef({ fs: this._fsModule, gitdir: this._gitdir, ref: refName });
      return true;
    } catch {
      return false;
    }
  }

  /**
   * List all branch or tag names.
   *
   * @returns Array of ref names (without the refs/heads/ or refs/tags/ prefix).
   */
  async list(): Promise<string[]> {
    if (this._isTags) {
      return git.listTags({ fs: this._fsModule, gitdir: this._gitdir });
    }
    return git.listBranches({ fs: this._fsModule, gitdir: this._gitdir });
  }

  /**
   * Async iteration over branch/tag names.
   */
  async *[Symbol.asyncIterator](): AsyncGenerator<string> {
    const names = await this.list();
    for (const name of names) {
      yield name;
    }
  }

  /**
   * Get the current (HEAD) branch name, or null if HEAD is dangling.
   *
   * Only valid for branches; throws for tags. Cheap -- does not
   * construct an FS object.
   *
   * @returns Branch name string, or null.
   * @throws {Error} If called on tags.
   */
  async getCurrentName(): Promise<string | null> {
    if (this._isTags) throw new Error('Tags do not have a current branch');
    const branch = await git.currentBranch({
      fs: this._fsModule,
      gitdir: this._gitdir,
      fullname: false,
    });
    return branch ?? null;
  }

  /**
   * Get the FS for the current (HEAD) branch, or null if HEAD is dangling.
   *
   * Only valid for branches; throws for tags.
   *
   * @returns Writable FS snapshot, or null.
   * @throws {Error} If called on tags.
   */
  async getCurrent(): Promise<FS | null> {
    if (this._isTags) throw new Error('Tags do not have a current branch');
    const name = await this.getCurrentName();
    if (!name) return null;
    try {
      return await this.get(name);
    } catch {
      return null;
    }
  }

  /**
   * Set the current branch (HEAD symbolic ref target).
   *
   * Only valid for branches; throws for tags.
   *
   * @param name - Branch name to set as current.
   * @throws {Error} If called on tags.
   * @throws {Error} If the branch does not exist.
   */
  async setCurrent(name: string): Promise<void> {
    if (this._isTags) throw new Error('Tags do not have a current branch');
    if (!(await this.has(name))) throw new Error(`Branch not found: '${name}'`);
    await git.writeRef({
      fs: this._fsModule,
      gitdir: this._gitdir,
      ref: 'HEAD',
      value: `refs/heads/${name}`,
      symbolic: true,
      force: true,
    });
  }

  /**
   * Read reflog entries for a branch.
   *
   * @param name - Branch name (e.g. 'main').
   * @returns List of ReflogEntry objects, newest first.
   * @throws {Error} If called on tags.
   * @throws {Error} If the branch does not exist.
   */
  async reflog(name: string): Promise<ReflogEntry[]> {
    if (this._isTags) throw new Error('Tags do not have reflog');
    if (!(await this.has(name))) throw new Error(`Key not found: ${name}`);
    return readReflog(this._fsModule, this._gitdir, name);
  }
}

File: rs/src/refdict.rs
use std::sync::Arc;

use crate::error::{Error, Result};
use crate::fs::Fs;
use crate::store::GitStore;
use crate::types::ReflogEntry;

/// A transient, borrowed view over a set of git references sharing a common
/// prefix (e.g. `refs/heads/` or `refs/tags/`).
///
/// `store.branches()` and `store.tags()` both return `RefDict` instances.
/// Branches yield writable [`Fs`] snapshots; tags yield read-only ones.
pub struct RefDict<'a> {
    store: &'a GitStore,
    prefix: &'static str,
}

impl<'a> RefDict<'a> {
    pub(crate) fn new(store: &'a GitStore, prefix: &'static str) -> Self {
        Self { store, prefix }
    }

    fn full_name(&self, name: &str) -> String {
        format!("{}{}", self.prefix, name)
    }

    /// Whether this RefDict is for branches (writable Fs) or tags (readonly).
    fn is_branch_prefix(&self) -> bool {
        self.prefix == "refs/heads/"
    }

    /// Whether this RefDict is for tags.
    fn is_tags(&self) -> bool {
        self.prefix == "refs/tags/"
    }

    /// Build an `Fs` from a resolved commit OID and ref name.
    fn fs_for_ref(&self, commit_oid: gix::ObjectId, name: &str) -> Result<Fs> {
        let writable = self.is_branch_prefix();
        Fs::from_commit(Arc::clone(&self.store.inner), commit_oid, Some(name.to_string()), Some(writable))
    }

    /// Get the [`Fs`] snapshot for the named branch or tag.
    ///
    /// Branches return a writable `Fs`; tags return a read-only `Fs`.
    ///
    /// # Errors
    /// Returns [`Error::NotFound`] if the ref does not exist.
    pub fn get(&self, name: &str) -> Result<Fs> {
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;
        let refname = self.full_name(name);
        let reference = repo
            .find_reference(refname.as_str())
            .map_err(|_| Error::not_found(format!("ref '{}' not found", name)))?;
        let commit_oid = reference.id().detach();
        drop(repo);
        self.fs_for_ref(commit_oid, name)
    }

    /// Point the named ref at the commit of `fs`.
    ///
    /// For branches, updates (or creates) the ref and writes a reflog entry.
    /// For tags, creates the ref but returns [`Error::KeyExists`] if the tag
    /// already exists (tags are immutable).
    ///
    /// # Arguments
    /// * `name` - Branch or tag name (e.g. `"main"`).
    /// * `fs` - The snapshot whose commit will become the ref target.
    ///
    /// # Errors
    /// * [`Error::InvalidRefName`] if `name` is not a valid git ref name.
    /// * [`Error::InvalidPath`] if `fs` belongs to a different repository.
    /// * [`Error::KeyExists`] if setting a tag that already exists.
    pub fn set(&self, name: &str, fs: &Fs) -> Result<()> {
        // 1. Validate ref name
        crate::paths::validate_ref_name(name)?;

        // 2. Same-repo check
        let same = Arc::ptr_eq(&self.store.inner, &fs.inner) || {
            let self_canon = std::fs::canonicalize(&self.store.inner.path).ok();
            let fs_canon = std::fs::canonicalize(&fs.inner.path).ok();
            self_canon.is_some() && self_canon == fs_canon
        };
        if !same {
            return Err(Error::invalid_path(
                "Fs belongs to a different repository".to_string(),
            ));
        }

        let commit_oid = fs
            .commit_oid
            .ok_or_else(|| Error::git_msg("Fs has no commit".to_string()))?;
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;
        let refname = self.full_name(name);

        // 3. Tag overwrite protection
        if self.is_tags() {
            if repo.find_reference(refname.as_str()).is_ok() {
                return Err(Error::key_exists(format!("tag '{}' already exists", name)));
            }
        }

        // Read old OID for reflog
        let old_oid = repo
            .find_reference(refname.as_str())
            .ok()
            .map(|r| r.id().detach());

        use gix::refs::transaction::PreviousValue;
        repo.reference(refname.as_str(), commit_oid, PreviousValue::Any, "refdict: set")
            .map_err(Error::git)?;

        // Write reflog entry
        let old_sha = old_oid
            .map(|o| format!("{}", o))
            .unwrap_or_else(|| crate::reflog::ZERO_SHA.to_string());
        let now = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap_or_default();
        let _ = crate::reflog::write_reflog_entry(
            &self.store.inner.path,
            &refname,
            &ReflogEntry {
                old_sha,
                new_sha: format!("{}", commit_oid),
                committer: format!(
                    "{} <{}>",
                    self.store.inner.signature.name, self.store.inner.signature.email
                ),
                timestamp: now.as_secs(),
                message: format!("refdict: set {}", name),
            },
        );

        Ok(())
    }

    /// Point the named ref at the commit of `fs` and return a new writable [`Fs`]
    /// bound to the updated ref.
    ///
    /// Convenience wrapper: equivalent to calling [`set`](Self::set) followed by
    /// [`get`](Self::get).
    pub fn set_to(&self, name: &str, fs: &Fs) -> Result<Fs> {
        self.set(name, fs)?;
        self.get(name)
    }

    /// Point the named ref at the commit of `fs` and return the previous [`Fs`]
    /// (or `None` if the ref did not exist before).
    pub fn set_and_get(&self, name: &str, fs: &Fs) -> Result<Option<Fs>> {
        let old = self.try_get(name)?;
        self.set(name, fs)?;
        Ok(old)
    }

    /// Delete the named branch or tag.
    ///
    /// # Errors
    /// Returns a git error if the ref does not exist or cannot be deleted.
    pub fn delete(&self, name: &str) -> Result<()> {
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;
        let refname = self.full_name(name);

        use gix::refs::transaction::{Change, PreviousValue, RefEdit, RefLog};
        use gix::refs::FullName;

        let edit = RefEdit {
            change: Change::Delete {
                expected: PreviousValue::Any,
                log: RefLog::AndReference,
            },
            name: FullName::try_from(refname).map_err(|e| Error::git(e))?,
            deref: false,
        };
        repo.edit_reference(edit).map_err(Error::git)?;
        Ok(())
    }

    /// Returns `true` if the named branch or tag exists.
    pub fn has(&self, name: &str) -> Result<bool> {
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;
        let refname = self.full_name(name);
        Ok(repo.find_reference(refname.as_str()).is_ok())
    }

    /// List all ref short names under this prefix, sorted alphabetically.
    ///
    /// For branches this returns names like `["dev", "main"]`; for tags
    /// `["v1.0", "v2.0"]`.
    pub fn list(&self) -> Result<Vec<String>> {
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;

        let refs_platform = repo.references().map_err(Error::git)?;
        let mut names = Vec::new();
        for r in refs_platform.prefixed(self.prefix).map_err(Error::git)? {
            if let Ok(reference) = r {
                let full_name = reference.name().as_bstr().to_string();
                if let Some(short) = full_name.strip_prefix(self.prefix) {
                    names.push(short.to_string());
                }
            }
        }
        names.sort();
        Ok(names)
    }

    /// Return all `(name, Fs)` pairs, sorted by name.
    pub fn iter(&self) -> Result<Vec<(String, Fs)>> {
        let pairs = {
            let repo = self
                .store
                .inner
                .repo
                .lock()
                .map_err(|e| Error::git_msg(e.to_string()))?;

            let refs_platform = repo.references().map_err(Error::git)?;
            let mut raw_pairs = Vec::new();
            for r in refs_platform.prefixed(self.prefix).map_err(Error::git)? {
                if let Ok(reference) = r {
                    let full_name = reference.name().as_bstr().to_string();
                    if let Some(short) = full_name.strip_prefix(self.prefix) {
                        let oid = reference.id().detach();
                        raw_pairs.push((short.to_string(), oid));
                    }
                }
            }
            raw_pairs.sort_by(|a, b| a.0.cmp(&b.0));
            raw_pairs
        };

        let mut result = Vec::with_capacity(pairs.len());
        for (name, oid) in pairs {
            let fs = self.fs_for_ref(oid, &name)?;
            result.push((name, fs));
        }
        Ok(result)
    }

    /// Get the current branch name (HEAD symbolic target within this prefix).
    ///
    /// Returns `Ok(None)` when HEAD is detached or dangling.
    /// Always returns `Ok(None)` for tags (tags do not have a "current" concept).
    /// This is a cheap operation — it does not construct an [`Fs`] object.
    pub fn get_current_name(&self) -> Result<Option<String>> {
        if self.is_tags() {
            return Ok(None);
        }
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;

        match repo.find_reference("HEAD") {
            Ok(head) => match head.target().try_name() {
                Some(name) => {
                    let name_str = name.as_bstr().to_string();
                    Ok(name_str.strip_prefix(self.prefix).map(|s| s.to_string()))
                }
                None => Ok(None),
            },
            Err(_) => Ok(None),
        }
    }

    /// Get the [`Fs`] for the current (HEAD) branch.
    ///
    /// Returns `Ok(None)` if HEAD is dangling or detached, or for tags.
    pub fn get_current(&self) -> Result<Option<Fs>> {
        if self.is_tags() {
            return Ok(None);
        }
        match self.get_current_name()? {
            Some(name) => match self.get(&name) {
                Ok(fs) => Ok(Some(fs)),
                Err(_) => Ok(None),
            },
            None => Ok(None),
        }
    }

    /// Set the current branch (HEAD symbolic ref target).
    ///
    /// # Errors
    /// Returns [`Error::Permission`] for tags (tags have no current branch).
    pub fn set_current(&self, name: &str) -> Result<()> {
        if self.is_tags() {
            return Err(Error::permission("tags do not support set_current"));
        }
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;

        let target_refname = self.full_name(name);

        use gix::refs::transaction::{Change, LogChange, PreviousValue, RefEdit, RefLog};
        use gix::refs::{FullName, Target};

        let edit = RefEdit {
            change: Change::Update {
                log: LogChange {
                    mode: RefLog::AndReference,
                    force_create_reflog: false,
                    message: format!("set current: {}", name).into(),
                },
                expected: PreviousValue::Any,
                new: Target::Symbolic(
                    FullName::try_from(target_refname).map_err(|e| Error::git(e))?,
                ),
            },
            name: FullName::try_from("HEAD".to_string()).map_err(|e| Error::git(e))?,
            deref: false,
        };
        repo.edit_reference(edit).map_err(Error::git)?;
        Ok(())
    }

    /// Read the reflog entries for the named branch.
    ///
    /// Returns a list of [`ReflogEntry`] objects recording each branch movement.
    ///
    /// # Errors
    /// * [`Error::Permission`] for tags (tags do not have reflogs).
    /// * [`Error::NotFound`] if no reflog file exists for the branch.
    pub fn reflog(&self, name: &str) -> Result<Vec<ReflogEntry>> {
        if self.is_tags() {
            return Err(Error::permission("tags do not have reflogs"));
        }
        let refname = self.full_name(name);
        crate::reflog::read_reflog(&self.store.inner.path, &refname)
    }

    /// Internal: get ref as Option (for set_and_get).
    fn try_get(&self, name: &str) -> Result<Option<Fs>> {
        let repo = self
            .store
            .inner
            .repo
            .lock()
            .map_err(|e| Error::git_msg(e.to_string()))?;
        let refname = self.full_name(name);
        match repo.find_reference(refname.as_str()) {
            Ok(reference) => {
                let commit_oid = reference.id().detach();
                drop(repo);
                Ok(Some(self.fs_for_ref(commit_oid, name)?))
            }
            Err(_) => Ok(None),
        }
    }
}
